test/manipulation.test.js

```diff
@@ -1686,7 +1700,73 @@ describe('manipulation', function() {
     });
 
     // TODO: implement destroy with filtered set
-    it('should destroy filtered set of records');
+    it.skip('should destroy filtered set of records');
+
+    it('should not delete using condition from filter', function(done) {
+      Person.destroyAll({where: {name: 'John'}}, function(err, info) {
+        should.exist(err);
+        err.should.match(/Filter object detected/);
+        Person.find({fields: 'name'}, function(err, users) {
+          should.not.exist(err);
+          users.should.containDeep([
+            {name: 'Jane'},
+            {name: 'Mary'},
+            {name: 'John'},
+          ]);
+          done();
+        });
+      });
+    });
+
+    it('should not delete all for bad and condition', function(done) {
+      Person.destroyAll({and: [{name: 'Jane'}, {where: {name: 'John'}}]}, function(err, info) {
+        should.not.exist(err);
+        Person.find({fields: 'name'}, function(err, users) {
+          should.not.exist(err);
+          let userObjects = [];
+          users.forEach(user => userObjects.push(user.toObject()));
+          userObjects.should.containDeep([
+            {name: 'John'},
+            {name: 'Jane'},
+            {name: 'Mary'},
+          ]);
+          done();
+        });
+      });
+    });
+
+    it('should delete all for undefined object', function(done) {
+      Person.destroyAll(undefined, function(err, info) {
+        should.not.exist(err);
+        Person.find({fields: 'name'}, function(err, users) {
+          should.not.exist(err);
+          users.should.containDeep([]);
+          done();
+        });
+      });
+    });
+
+    it('should delete all for empty object', function(done) {
+      Person.destroyAll({}, function(err, info) {
```

Comment:
I think this is the dangerous behaviour we want to eliminate. If a user wants to destroyAll, then it should be called as such (with `undefined` or nothin) but if an object is present then it should be a filter and not an empty object IMO.

Reply:
Wouldn't such change be backwards incompatible and require a semver-major release?

Reply:
Yes. This is a good discussion point. I agree with changing the usage to use `undefined` value for the `where` object (which is also the case when it's omitted in API Explorer since it's optional) to delete all records. That means changing our default assignment to `undefined` since in `dao` we default to an empty object. Then, we need to make sure that the connector specific methods also adhere to this behaviour. I also want to implement rejection of empty objects and to advise users to use `undefined` instead. Yes that would mean a semver-major release, but perhaps we can mitigate it by introducing a flag for this new behaviour as well? Please keep in mind that this is also what the users complained about in the [original issue](https://github.com/strongloop/loopback/issues/3094).

Reply:
Please note that detecting `{}` can be tricky.

Should we treat `{userId: undefined}` the same way as `{}`? Or does the first example mean "match all records that have no `userId` set"? With SQL connectors, this can be probably implemented as a NULL check, but I don't know what's the level of support in noSQL connectors for such query.

> perhaps we can mitigate it by introducing a flag for this new behaviour as well?

Yeah, we will have to.

> Please keep in mind that this is also what the users complained about in the original issue.

As I understand the discussion in the original issue, the biggest pain point is the situation when people accidentally wrap the condition inside `where`, which is an extension of the situation when they make a typo in a property name.

The case of `where={}` is an edge case that was brought up later, as we started to discuss possible fixes.

I would personally focus on the first part (i.e. fix connectors to stop ignoring unknown properties) and leave changes in handling of the empty `where` object for later.

Baby steps FTW!

Reply:
> Should we treat {userId: undefined} the same way as {}?

I don't think that `userId: undefined` is the same as an empty object. I think it makes more sense for that to be treated as `match all records that have no userId set` while an empty object should not be allowed. If the intent is to delete all object then the user should call `destroyAll()` with no arguments. 

Any reason we can't have a breaking change with this module? I wouldn't consider this a breaking change considering I think this was a design oversight that we are fixing but to adhere to semver we can do a major release (at least that's my opinion).

Reply:
I agree with making fixes in small increments. I'll remove this test case for now (thoughts?). We could definitely follow up with issues regarding how we would like to handle empty objects. I agree that `{userId: undefined}` shouldn't be the same as `{}` but I think we should either reject it or yes treat it as `match all records that have no userId set`.